components/core/src/crypto/keys/cache.rs

```diff
@@ -0,0 +1,600 @@
+    // kind of key we're dealing with. For example "service.pub" and
+    // "service.box.key" for service encryption keys, "sig.pub"
+    // and "sig.key" for signing keys, etc. This would, of course,
+    // have to be done in a backwards-compatible way for all the keys
+    // currently in existence.
+    pub fn write_key<K>(&self, key: &K) -> Result<()>
+        where K: KeyFile
+    {
+        let keyfile = self.path_in_cache(key);
+        let content = key.to_key_string();
+
+        if keyfile.is_file() {
+            let existing_hash = hash::hash_file(&keyfile)?;
+            let new_hash = hash::hash_string(&content);
+            if existing_hash != new_hash {
+                let msg = format!("Existing key file {} found but new version hash is different, \
+                                   failing to write new file over existing. (existing = {}, \
+                                   incoming = {})",
+                                  keyfile.display(),
+                                  existing_hash,
+                                  new_hash);
+                return Err(Error::CryptoError(msg));
+            }
+        } else {
+            // Technically speaking, this probably doesn't really need
+            // to be an atomic write process, since we just tested
+            // that the file doesn't currently exist. It does,
+            // however, bundle up writing with platform-independent
+            // permission setting, which is *super* convenient.
+            let w = AtomicWriter::new_with_permissions(&keyfile, K::permissions())?;
+            w.with_writer(|f| f.write_all(content.as_ref()))?;
+        }
+        Ok(())
+    }
+
+    /// Note: name is just the name, not the name + revision
+    pub fn latest_ring_key_revision(&self, name: &str) -> Result<RingKey> {
+        self.fetch_latest_revision::<RingKey>(name)
+    }
+
+    pub fn latest_secret_origin_signing_key(&self, name: &str) -> Result<SecretOriginSigningKey> {
+        self.fetch_latest_revision::<SecretOriginSigningKey>(name)
+    }
+
+    pub fn latest_public_origin_signing_key(&self, name: &str) -> Result<PublicOriginSigningKey> {
+        self.fetch_latest_revision::<PublicOriginSigningKey>(name)
+    }
+
+    pub fn latest_user_secret_key(&self, user_name: &str) -> Result<UserSecretEncryptionKey> {
+        self.fetch_latest_revision::<UserSecretEncryptionKey>(user_name)
+    }
+
+    pub fn latest_origin_public_encryption_key(&self,
+                                               name: &str)
+                                               -> Result<OriginPublicEncryptionKey> {
+        self.fetch_latest_revision::<OriginPublicEncryptionKey>(name)
+    }
+
+    /// Name should be in `"service.group@org"` format.
+    pub fn latest_service_public_key(&self, name: &str) -> Result<ServicePublicEncryptionKey> {
+        self.fetch_latest_revision::<ServicePublicEncryptionKey>(name)
+    }
```

Comment:
Not necessarily for this PR, but as a follow on, it might be nice to expose these `fetch_latest_*` to the CLI key commands so that an operator can rely on Habitat to do the lifting and determine the latest key. Otherwise, speaking from experience, it can be bit cumbersome in automation to generate a key pair then inspect the filesystem to determine the latest version and name of the key file.

Reply:
Good idea!